Refactor C++ API parity test mechanism and turn it on in CI again

yf225 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#35190 Refactor C++ API parity test mechanism and turn it on in CI again**
* #35189 Move NewModuleTest and NewCriterionTest from test_nn.py to common_nn.py

The following are the main changes:
- The main logic of C++ API parity test mechanism is moved from `test/test_cpp_api_parity.py` to `test/cpp_api_parity/module_impl_check.py` and `test/cpp_api_parity/functional_impl_check.py`, so that there is a clear separation between module tests and functional tests, although they still share a lot of common utility functions which are all in `test/cpp_api_parity/utils.py`.
- `cpp_constructor_args` / `cpp_options_args` / `cpp_function_call` are added as appropriate to all test params dict in `torch/testing/_internal/common_nn.py`, to indicate how to run C++ API parity test for that test params dict.

Follow-up items:
- Functional backwards parity test (i.e. testing whether the C++ inputs receive the same gradients as the Python inputs) will be added in the future.
- Module init test (i.e. testing whether C++ module accepts the same constructor options as the corresponding Python module) is removed and will be added again in the future.
- There's a lot of I/O going on with these tests, which increases the risk of flakiness. If that becomes a problem in the future maybe someone can use pybind to connect the result of the cpp api to python.
- the C++ API parity test mechanism re-implements running the forward + backward of a module test and reimplements running the forward of a functional test. The Module tests in test_nn.py still run using their own, separate mechanism. Is there a plan in the future to unify the two mechanisms? Afaict the C++ API parity test mechanism doesn't support some things that the test_nn.py mechanism does; off of the top of my head check_gradgrad (which defaults to True) or precision checks (check_half) don't look supported

**Test Plan:**
Make sure `test/test_cpp_api_parity.py` is running in the CI.

Differential Revision: [D20588198](https://our.internmc.facebook.com/intern/diff/D20588198)

Commit:
Turn on C++ API parity test again

Commit:
Update on "Turn on C++ API parity test again"

Turn on C++ API parity test again

gh-metadata: pytorch pytorch 35190 gh/yf225/106/head

Commit:
Update on "Turn on C++ API parity test again"

Turn on C++ API parity test again

gh-metadata: pytorch pytorch 35190 gh/yf225/106/head

Commit:
Update on "Turn on C++ API parity test again"

Turn on C++ API parity test again

gh-metadata: pytorch pytorch 35190 gh/yf225/106/head

Commit:
Update on "Turn on C++ API parity test again"

Turn on C++ API parity test again

gh-metadata: pytorch pytorch 35190 gh/yf225/106/head

Commit:
Update on "Turn on C++ API parity test again"

Turn on C++ API parity test again

gh-metadata: pytorch pytorch 35190 gh/yf225/106/head

Commit:
Update on "Turn on C++ API parity test again"

Turn on C++ API parity test again

gh-metadata: pytorch pytorch 35190 gh/yf225/106/head

Commit:
Update on "Turn on C++ API parity test again"

Turn on C++ API parity test again

gh-metadata: pytorch pytorch 35190 gh/yf225/106/head

Commit:
Update on "Turn on C++ API parity test again"

Turn on C++ API parity test again

gh-metadata: pytorch pytorch 35190 gh/yf225/106/head

Commit:
Update on "Turn on C++ API parity test again"

Turn on C++ API parity test again

gh-metadata: pytorch pytorch 35190 gh/yf225/106/head

Commit:
Update on "Turn on C++ API parity test again"

Turn on C++ API parity test again

gh-metadata: pytorch pytorch 35190 gh/yf225/106/head

Commit:
Update on "Refactor C++ API parity test mechanism and turn it on in CI again"

Turn on C++ API parity test again

gh-metadata: pytorch pytorch 35190 gh/yf225/106/head

Commit:
Update on "Refactor C++ API parity test mechanism and turn it on in CI again"

Turn on C++ API parity test again

gh-metadata: pytorch pytorch 35190 gh/yf225/106/head

Commit:
Update on "Refactor C++ API parity test mechanism and turn it on in CI again"

Turn on C++ API parity test again

gh-metadata: pytorch pytorch 35190 gh/yf225/106/head

Commit:
Update on "Refactor C++ API parity test mechanism and turn it on in CI again"

Turn on C++ API parity test again

gh-metadata: pytorch pytorch 35190 gh/yf225/106/head

Commit:
Update on "Refactor C++ API parity test mechanism and turn it on in CI again"

Turn on C++ API parity test again

gh-metadata: pytorch pytorch 35190 gh/yf225/106/head

zou3519 commented on test/test_cpp_api_parity.py:
```diff
@@ -1,687 +1,64 @@
+which is tracked at https://github.com/pytorch/pytorch/issues/25883.)
+
+However, if any of the above is proven to be too complicated, you can just add
+`test_cpp_api_parity=False` to any failing test params dict in `torch/testing/_internal/common_nn.py`,
```
Maybe also ask them to file an issue?

zou3519 commented on test/test_cpp_api_parity.py:
```diff
@@ -1,687 +1,64 @@
-    if has_test(test_name):
-        raise RuntimeError("Found two tests with the same name: " + test_name)
-    setattr(TestCppApiParity, test_name, test_fn)
+print_cpp_source = True
```
Are you sure we want this by default? It fills up the logs a lot

zou3519 commented on test/test_cpp_api_parity.py:
```diff
@@ -1,687 +1,64 @@
+MESSAGE_HOW_TO_FIX_CPP_PARITY_TEST_FAILURE = '''
+What should I do when C++ API parity test is failing?
+
+- If you are changing the implementation of an existing `torch.nn` module / functional:
```
nit: Instead of functional, "`torch.nn.functional` function" or "`torch.nn.functional` api" is clearer

zou3519 commented on test/test_cpp_api_parity.py:
```diff
@@ -1,687 +1,64 @@
+Answer: Ideally you should also change the C++ API implementation for that module / functional
+(you can start by searching for the module / functional name in `torch/csrc/api/` folder).
+
+- If you are adding a new test params dict for an existing `torch.nn` module / functional:
```
What is a "new test params dict"?

zou3519 replied:
Oh, I see. Maybe call them "ModuleTest" or NN module test or link the dicts in common_nn.py?

zou3519 commented on test/test_cpp_api_parity.py:
```diff
@@ -1,687 +1,64 @@
 
 if __name__ == "__main__":
-    common.run_tests()
+  common.run_tests()
```
We should do 4 space indentation to match the rest of the project

zou3519 commented on test/test_cpp_api_parity.py:
```diff
@@ -1,687 +1,64 @@
+  len(sample_functional.functional_tests) * len(devices)
 
+module_impl_check.build_cpp_tests(TestCppApiParity, print_cpp_source=print_cpp_source)
+functional_impl_check.build_cpp_tests(TestCppApiParity, print_cpp_source=print_cpp_source)
```
The building should happen in the `if __name__ == __main__` block as well. Let's say that someone wants to check what tests TestCppApiParity has. Then we don't want to force them to build the cpp_tests before checking what tests are available.

zou3519 commented on test/test_cpp_api_parity.py:
```diff
@@ -1,687 +1,64 @@
-                    module_full_name, parity_table_path))
-
-        has_impl_parity, _ = parity_table['torch.nn'][module_full_name]
+parity_table_path = os.path.join(os.path.dirname(__file__), 'cpp_api_parity', 'parity-tracker.md')
```
nit: constants should be all caps

zou3519 commented on torch/testing/_internal/common_nn.py:
```diff
@@ -274,9 +294,12 @@ def _rand_tensor_non_equal(*size):
 
 def wrap_functional(fn, **kwargs):
-    class FunctionalModule(nn.Module):
+    class FunctionalModule(torch.nn.Module):
```
Why did this have to change?

zou3519 commented on torch/testing/_internal/common_nn.py:
```diff
@@ -3342,9 +3699,11 @@ def padding3d_circular(input, pad):
     dict(
         module_name='BCELoss',
-        constructor_args_fn=lambda: (torch.rand(()),),
+        constructor_args_fn=lambda: (criterion_weight_rand_zero_dim,),
```
Why did this change?

yf225 replied:
removed :D

zou3519 commented on torch/testing/_internal/common_nn.py:
```diff
@@ -3370,9 +3730,11 @@ def padding3d_circular(input, pad):
     dict(
         module_name='MultiLabelSoftMarginLoss',
-        constructor_args=(torch.rand(10),),
+        constructor_args=(criterion_weight_rand_10,),
```
Why did this change?

yf225 replied:
removed :D

zou3519 commented on torch/testing/_internal/common_nn.py:
```diff
@@ -2940,11 +3263,18 @@ def padding3d_circular(input, pad):
     'CTCLoss': ctcloss_reference,
 }
 
+# NOTE: Instead of creating these weight tensors on the spot during
```
Is this sufficient? Are there any modules that create random weights?

yf225 replied:
Thanks for the catch - I actually don't think these constant tensors are necessary now, because we are always copying Python module's params / buffers to the C++ module. I have removed them :D

zou3519 commented on test/cpp_api_parity/torch_nn_modules.py:
```diff
@@ -1,193 +1,12 @@
+TorchNNModuleMetadata = namedtuple(
+    'TorchNNModuleMetadata',
+    [
+        'cpp_sources',
```
When does one need to provide a `cpp_sources`?

yf225 replied:
We only need to provide it for `SampleModule` and `sample_functional` (since their C++ implementation is not available in libtorch)

zou3519 replied:
Do you see us adding anything to this named tuple in the future? We really don't need a named tuple if there is only going to be a single argument.

Also, it would be nice to put your comment about that into the code (we only need to provide cpp_sources if we're testing an out-of-tree module and this is only done to test the testing harness)

yf225 replied:
I changed `module_metadata_map` to `module_cpp_sources` and make it just be a dict of module name -> cpp sources (same for `functional_metadata_map` as well), and also added a comment for when we need to provide cpp sources :D

zou3519 commented on test/cpp_api_parity/parity_table_parser.py:
```diff
@@ -0,0 +1,58 @@
+ParityStatus = namedtuple('ParityStatus', ['has_impl_parity', 'has_doc_parity'])
+
+'''
+This function expects the parity tracker Markdown file to have the following format:
```
It looks like the contents of this file were moved from `test/cpp_api_parity/__init__.py` -- did anything change?

yf225 replied:
Nope nothing changes :)

zou3519 reviewed (commented):
Added some initial questions, I am going to take a break for now

Commit:
Update on "Refactor C++ API parity test mechanism and turn it on in CI again"

Turn on C++ API parity test again

gh-metadata: pytorch pytorch 35190 gh/yf225/106/head

Commit:
Update on "Refactor C++ API parity test mechanism and turn it on in CI again"

Turn on C++ API parity test again

gh-metadata: pytorch pytorch 35190 gh/yf225/106/head

Commit:
Update on "Refactor C++ API parity test mechanism and turn it on in CI again"

Turn on C++ API parity test again

gh-metadata: pytorch pytorch 35190 gh/yf225/106/head

Commit:
Update on "Refactor C++ API parity test mechanism and turn it on in CI again"

Turn on C++ API parity test again

gh-metadata: pytorch pytorch 35190 gh/yf225/106/head

zou3519 commented on torch/testing/_internal/common_nn.py:
```diff
@@ -275,8 +295,11 @@ def _rand_tensor_non_equal(*size):
 
 def wrap_functional(fn, **kwargs):
     class FunctionalModule(nn.Module):
+        def __init__(self):
```
Why did the implementation of `wrap_functional` change? Is there a reason why the old one doesn't work?

yf225 replied:
Thanks for the catch - it was an artifact of an old implementation to get name of the functional from `.fn`, but it's not needed anymore :D

Commit:
Update on "Refactor C++ API parity test mechanism and turn it on in CI again"

Turn on C++ API parity test again

gh-metadata: pytorch pytorch 35190 gh/yf225/106/head

Commit:
Update on "Refactor C++ API parity test mechanism and turn it on in CI again"

Turn on C++ API parity test again

gh-metadata: pytorch pytorch 35190 gh/yf225/106/head

Commit:
Update on "Refactor C++ API parity test mechanism and turn it on in CI again"

Turn on C++ API parity test again

gh-metadata: pytorch pytorch 35190 gh/yf225/106/head

Commit:
Update on "Refactor C++ API parity test mechanism and turn it on in CI again"

Turn on C++ API parity test again

gh-metadata: pytorch pytorch 35190 gh/yf225/106/head

Commit:
Update on "Refactor C++ API parity test mechanism and turn it on in CI again"

Turn on C++ API parity test again

gh-metadata: pytorch pytorch 35190 gh/yf225/106/head

Commit:
Update on "Refactor C++ API parity test mechanism and turn it on in CI again"

Turn on C++ API parity test again

gh-metadata: pytorch pytorch 35190 gh/yf225/106/head

zou3519 commented on torch/testing/_internal/common_nn.py:
```diff
@@ -3410,7 +3752,8 @@ def padding3d_circular(input, pad):
         constructor_args=(0,),  # blank=0
-        extra_args=([50, 50, 50], [30, 25, 20]),  # input_lengths, target_lengths
+        cpp_constructor_args='torch::nn::CTCLossOptions().blank(0)',
+        extra_args=(torch.tensor([50, 50, 50]), torch.tensor([30, 25, 20])),  # input_lengths, target_lengths
```
This makes it so that we aren't testing that CTCLoss can take non-Tensor input_lengths. This could be a problem if someone unintentionally changes the API here.

I'm still trying to figure out how `extra_args` gets used by the cpp parity test so I have no suggestions here yet.

yf225 replied:
We always serialize `extra_args` as a list of tensors and load them into the cpp parity test, and changing it to also allow serializing lists of ints might be too ad-hoc. Maybe as a workaround, we can make a copy of this module test and use `extra_args=(torch.tensor([50, 50, 50]), torch.tensor([30, 25, 20]))`, and for the original one we put `test_cpp_api_parity=False`?

zou3519 replied:
Yeah, I think it would be good to add a copy of the module test as a workaround to make sure that works

yf225 replied:
Added :D

zou3519 commented on test/cpp_api_parity/sample_module.py:
```diff
@@ -1,124 +1,106 @@
     ),
     dict(
         fullname='SampleModule_no_parity',
         constructor=lambda: SampleModule(False, True),
```
Nit: this is easier to read if we pass False, True as keyword args: `lambda: SampleModule(has_parity=False, has_submodule=True)`. Otherwise the reader has to scroll up. Same for the next dict in this file

zou3519 commented on test/cpp_api_parity/sample_module.py:
```diff
@@ -1,124 +1,106 @@
         has_parity=False,
     ),
+    dict(
+        fullname='SampleModule_THIS_TEST_SHOULD_BE_SKIPPED',
```
What is being tested here? And is it enforced anywhere?

yf225 replied:
It is to test that setting the `test_cpp_api_parity=False` flag skips the C++ API parity test accordingly (otherwise this test would run and throw a parity error). I just added a comment in the code for this one :D

zou3519 replied:
Ah I see. Yeah, the comment is great, thanks for clarifying

Commit:
Update on "Refactor C++ API parity test mechanism and turn it on in CI again"

Turn on C++ API parity test again

gh-metadata: pytorch pytorch 35190 gh/yf225/106/head

Commit:
Update on "Refactor C++ API parity test mechanism and turn it on in CI again"

Turn on C++ API parity test again

gh-metadata: pytorch pytorch 35190 gh/yf225/106/head

zou3519 commented on test/cpp_api_parity/module_impl_check.py:
```diff
@@ -0,0 +1,289 @@
+
+# Build all C++ tests together, instead of once per test.
+def build_cpp_tests(unit_test_class, print_cpp_source=False):
+    if len(torch_nn_test_params_map) > 0:
```
When would this be equal to 0?

yf225 replied:
It should never be equal to 0 - I changed it to assert :D

zou3519 commented on test/cpp_api_parity/utils.py:
```diff
@@ -0,0 +1,225 @@
+
+    return test_fn
+
+extra_msg_on_failure = ''
```
What is this used for?

yf225 replied:
I used to put `MESSAGE_HOW_TO_FIX_CPP_PARITY_TEST_FAILURE` in `test/test_cpp_api_parity.py` and then inject it here, but I think it's bad code structure and just moved `MESSAGE_HOW_TO_FIX_CPP_PARITY_TEST_FAILURE` to this file instead :D

zou3519 commented on test/cpp_api_parity/module_impl_check.py:
```diff
@@ -0,0 +1,289 @@
+  c10::Dict<std::string, torch::Tensor> grad_dict;
+  for (const auto& param : module->named_parameters()) {
+    torch::Tensor grad = param.value().grad();
+    if (grad.is_sparse()) {
```
This doesn't really help us check parity between if the python implementation is giving sparse gradients and if the c++ implementation is giving sparse gradients right?

We don't have a lot of modules that do this (is embedding the only one?) so this isn't blocking but it would be good to note somewhere

yf225 replied:
Thanks for the catch! Fixed it by serializing `.indices()` and `.values()` separately for sparse tensors, so that we are able to check parity between if the python implementation is giving sparse gradients and if the c++ implementation is giving sparse gradients

zou3519 replied:
nice! 😅

Commit:
Update on "Refactor C++ API parity test mechanism and turn it on in CI again"

Turn on C++ API parity test again

gh-metadata: pytorch pytorch 35190 gh/yf225/106/head

glaringlee commented on test/cpp_api_parity/functional_impl_check.py:
```diff
@@ -0,0 +1,232 @@
+    run_cpp_test_fn_and_check_output()
+
+    # Remove temporary folder that stores C++ outputs
+    shutil.rmtree(cpp_tmp_folder)
```
This will raise an exception on error, should we add a comments here say this will raise exception on error?  or catch it?

yf225 replied:
Thanks for the catch! I added an assert above this line to check that `cpp_tmp_folder` always exists on disk (since the current form of the test mechanism does expect it to exist) :D

glaringlee replied:
What I am concern is that since this is to delete from hdd/sdd file system, it is not always reliable depending on how many instances currently using that machine and accessing the same disk, if your deletion request failed due to high disk traffic or some other temp connection issues, this shouldn't block the whole process. But I think facebook's cluster is well scaled generally, above issue should be catch in upper hardware layer, adding an assert here  to check folder existence is fine then.

zou3519 replied:
@glaringlee we run these tests on OSS as well (I'm not sure how many of these tests get run internally). If we're really scared of flakiness from `rmtree` not working, we could wrap the statement in a try block and don't catch any exceptions (lol):
```
try:
    shutil.rmtree(cpp_tmp_folder)
```

glaringlee replied:
@zou3519 @yf225, I am not sure how often this would happen, I do have experience that such failure happened one to two times a month on average. I agree that a single try would work just to make sure this wouldn't block the process even failed.

yf225 replied:
Thanks for the discussion! I added
```python
try:
    shutil.rmtree(cpp_tmp_folder)
except:
    pass
```
because otherwise Python is complaining about wrong indentation level at next line..

glaringlee commented on test/cpp_api_parity/module_impl_check.py:
```diff
@@ -0,0 +1,289 @@
+    const std::string& module_file_path,
+    const std::string& forward_output_file_path,
+    const std::string& backward_grad_dict_file_path) {
+  pybind11::gil_scoped_release no_gil;
```
Same comments here

glaringlee replied:
RAII guard same as above, not need to comments in this case.

glaringlee commented on test/cpp_api_parity/module_impl_check.py:
```diff
@@ -0,0 +1,289 @@
+    run_cpp_test_fn_and_check_output()
+
+    # Remove temporary folder that stores C++ outputs
+    shutil.rmtree(cpp_tmp_folder)
```
comment possible throwing exceptions here or catch it?

zou3519 commented on test/cpp_api_parity/functional_impl_check.py:
```diff
@@ -0,0 +1,232 @@
+# Expected substitutions:
+#
+# ${functional_variant_name}
+# ${cpp_tmp_folder}
```
I don't see this in the Template

yf225 replied:
Thanks for the catch! Fixed

zou3519 commented on test/cpp_api_parity/functional_impl_check.py:
```diff
@@ -0,0 +1,232 @@
+# 3. Run the forward pass on the C++ functional, and serialize the C++ functional's
+# forward output.
+#
+# 4. Compare Python/C++ functional's forward output. If they are the same, then we
```
Why do we need to check backwards in the NN parity test, but not in the functional parity test?

yf225 replied:
Thanks a lot for the catch! Backwards in functional parity test should also be checked, and I will add it to the list of follow-up items to do.

glaringlee reviewed (changes requested):
Some comments here, still reviewing the logic.

zou3519 commented on torch/testing/_internal/common_nn.py:
```diff
@@ -1849,248 +2065,288 @@ def fractional_max_pool3d_test(test_case):
     ),
     dict(
         constructor=wrap_functional(F.interpolate, size=12, scale_factor=None, mode='nearest'),
+        cpp_options_args='F::InterpolateFuncOptions().size(std::vector<int64_t>({12})).scale_factor(c10::nullopt).mode(torch::kNearest)',
```
just for my understanding: What happens in this case, where we have `cpp_options_args` but no `cpp_function_call`? What does the `F::interpolate` function call look like, do we pass the options directly to it?

yf225 replied:
That's right - in the case where we have `cpp_options_args` but no `cpp_function_call`, the C++ function call would look like: `F::interpolate(input, cpp_options_args)`.

In general, if we only have `cpp_options_args`, the function call is `F::functional_name(input, target, extra_args_1, extra_args_2, ..., cpp_options_args)`

yf225 replied:
I added an assert to check that there has to be one and only one of `cpp_options_args` and `cpp_function_call` that exists in the test dict :D

Commit:
Update on "Refactor C++ API parity test mechanism and turn it on in CI again"

Turn on C++ API parity test again

gh-metadata: pytorch pytorch 35190 gh/yf225/106/head

zou3519 commented on test/cpp_api_parity/functional_impl_check.py:
```diff
@@ -0,0 +1,232 @@
+
+# Build all C++ tests together, instead of once per test.
+def build_cpp_tests(unit_test_class, print_cpp_source=False):
+    if len(torch_nn_test_params_map) > 0:
```
this can't be == 0, right?

yf225 replied:
Fixed :D

Commit:
Update on "Refactor C++ API parity test mechanism and turn it on in CI again"

Turn on C++ API parity test again

gh-metadata: pytorch pytorch 35190 gh/yf225/106/head

zou3519 commented on test/cpp_api_parity/module_impl_check.py:
```diff
@@ -0,0 +1,293 @@
+#
+# ${module_variant_name}
+# ${module_qualified_name}
+# ${cpp_tmp_folder}
```
I don't see this in this Template either

yf225 replied:
Removed :D

zou3519 commented on test/cpp_api_parity/module_impl_check.py:
```diff
@@ -0,0 +1,293 @@
+# ${cpp_tmp_folder}
+# ${cpp_args_construction_stmts}
+# ${cpp_constructor_args}
+# ${device}
```
Do we have to use device as a template?

yf225 replied:
I think we have two options: 1. use device as a template, 2. pass in device as a function argument. The first option is probably better because we already encode device information (i.e. `cpu` vs `cuda`)  in `module_variant_name`, and so it is probably not a good idea to allow passing in a different device as function argument.

zou3519 replied:
Yup, that sounds reasonable, thanks for the explanation

zou3519 commented on test/cpp_api_parity/module_impl_check.py:
```diff
@@ -0,0 +1,293 @@
+# Expected substitutions:
+#
+# ${module_variant_name}
+# ${module_qualified_name}
```
What's the difference between this and `module_variant_name`?

yf225 replied:
`module_qualified_name` is something like `torch::nn::Linear` (with the `torch::nn::` qualifier), `module_variant_name` is something like `Linear_no_bias_cpu`. I just added some comments to explain this better :D

zou3519 replied:
thanks!

Commit:
Update on "Refactor C++ API parity test mechanism and turn it on in CI again"

Turn on C++ API parity test again

gh-metadata: pytorch pytorch 35190 gh/yf225/106/head

Commit:
Update on "Refactor C++ API parity test mechanism and turn it on in CI again"

Turn on C++ API parity test again

gh-metadata: pytorch pytorch 35190 gh/yf225/106/head

Commit:
Update on "Refactor C++ API parity test mechanism and turn it on in CI again"

Turn on C++ API parity test again

gh-metadata: pytorch pytorch 35190 gh/yf225/106/head

Commit:
Update on "Refactor C++ API parity test mechanism and turn it on in CI again"

Turn on C++ API parity test again

gh-metadata: pytorch pytorch 35190 gh/yf225/106/head

yf225 commented on test/cpp_api_parity/functional_impl_check.py:
```diff
@@ -0,0 +1,232 @@
+void ${functional_variant_name}_test_forward(
+    const std::string& arg_dict_file_path,
+    const std::string& forward_output_file_path) {
+  pybind11::gil_scoped_release no_gil;
```
hmm I think this is an RAII guard, and `no_gil` would be destroyed automatically (and we are back in GIL land) on exit of this scope? @albanD curious is this correct?

albanD replied:
Yes this is an RAII construct. This releases the GIL on construction and re-acquire it on destruction.

glaringlee replied:
Ah, I see.

Commit:
Update on "Refactor C++ API parity test mechanism and turn it on in CI again"

Turn on C++ API parity test again

gh-metadata: pytorch pytorch 35190 gh/yf225/106/head

Commit:
Update on "Refactor C++ API parity test mechanism and turn it on in CI again"

Turn on C++ API parity test again

gh-metadata: pytorch pytorch 35190 gh/yf225/106/head

Commit:
Update on "Refactor C++ API parity test mechanism and turn it on in CI again"

Turn on C++ API parity test again

gh-metadata: pytorch pytorch 35190 gh/yf225/106/head

Commit:
Update on "Refactor C++ API parity test mechanism and turn it on in CI again"

Turn on C++ API parity test again

gh-metadata: pytorch pytorch 35190 gh/yf225/106/head

Commit:
Update on "Refactor C++ API parity test mechanism and turn it on in CI again"

Turn on C++ API parity test again

gh-metadata: pytorch pytorch 35190 gh/yf225/106/head

Commit:
Update on "Refactor C++ API parity test mechanism and turn it on in CI again"

Turn on C++ API parity test again

gh-metadata: pytorch pytorch 35190 gh/yf225/106/head

Commit:
Update on "Refactor C++ API parity test mechanism and turn it on in CI again"

Turn on C++ API parity test again

gh-metadata: pytorch pytorch 35190 gh/yf225/106/head

zou3519 commented on test/test_cpp_api_parity.py:
```diff
@@ -1,687 +1,44 @@
-        raise RuntimeError("Found two tests with the same name: " + test_name)
-    setattr(TestCppApiParity, test_name, test_fn)
+# NOTE: turn this on if you want to print source code of all C++ tests (e.g. for debugging purpose)
+print_cpp_source = False
```
nit: constant should be all caps "PRINT_CPP_SOURCE"

zou3519 commented on test/cpp_api_parity/functional_impl_check.py:
```diff
@@ -0,0 +1,239 @@
+    inputs = set_python_tensors_requires_grad([arg_value for _, arg_value in test_params.arg_dict['input']])
+    inputs = inputs + [arg_value for _, arg_value in test_params.arg_dict['target']]
+    inputs = inputs + [arg_value for _, arg_value in test_params.arg_dict['extra_args']]
+    inputs = move_python_tensors_to_device(inputs, device)
```
Is it intentional that we're setting the requires grad to be True before we move the tensors onto the device?

yf225 replied:
Thanks a lot for the catch! No it's not intentional, and I just changed their order to always move to device first before setting requires grad

zou3519 commented on test/cpp_api_parity/functional_impl_check.py:
```diff
@@ -0,0 +1,239 @@
+def add_torch_nn_functional_impl_parity_tests(parity_table, unit_test_class, test_params_dicts, test_instance_class, devices):
+    for test_params_dict in test_params_dicts:
+        # Skip all `torch.nn` module tests, since they are handled by another test suite.
+        if 'FunctionalModule' not in str(test_params_dict.get('constructor', '')):
```
Should this be "wrap_functional"?

zou3519 replied:
If so, it doesn't seem robust to assume that all functional tests use `wrap_functional`. One alternative that is more robust but potentially more verbose is to add a "is_functional" flag to each dict...

yf225 replied:
Yeah I thought about that and felt that adding `is_functional` flag to all functional dicts might be a bit verbose.. Here is my updated approach:
1. Assert that the number of tests we have in C++ parity test be exactly equal to the number of NN module/functional tests we have in common_nn.py. This way, we can make sure we are always adding all NN module/functional tests into the C++ parity test.
2. We will always check whether `wrap_functional` exists in `str(test_params_dict.get('constructor', ''))` to identify whether to put this test in module parity test or functional parity test. Since this is "either-or", when we add a new test without `wrap_functional` we will route the test to module parity test, which will then throw an error if the module parity test mechanism cannot understand the newly added test, prompting the author to either fix the test or improve the parity test mechanism to accept the new test format.

zou3519 commented on test/cpp_api_parity/functional_impl_check.py:
```diff
@@ -0,0 +1,239 @@
+        cpp_tmp_folder=tempfile.mkdtemp(),
+    )
+
+torch_nn_test_params_map = {}
```
Global state is hard to reason about. Is it necessary for this to be global?

zou3519 replied:
One way to avoid this is to make some class object:
```
for test_param in ...:
      FunctionalImplCheck.add_test(blah)
FunctionalImplCheck.build_tests()
FunctionalImplCheck.add_tests_to(TestCppApiParity)
```
This is sort of what the code is doing right now, except the code is modifying the `functional_impl_check` module and essentially treating it like a class with some state (the torch_nn_test_params_map)

zou3519 replied:
That would also make some other things more clear, like:
- `functional_impl_check.add_tests` is very different from `utils.add_test`, but they have similar names

zou3519 replied:
But I'm not sure if this would actually make the code look nicer. Let me know your thoughts

yf225 replied:
Thanks for looking into it! I moved `torch_nn_test_params_map` into the `unit_test_class` object, so that there is no global state for `functional_impl_check` and `module_impl_check` anymore, but all methods can still access this global map object via `unit_test_class`

zou3519 commented on test/test_cpp_api_parity.py:
```diff
@@ -1,687 +1,44 @@
+    (common_nn.criterion_tests, common_nn.CriterionTest),
+    (common_nn.new_criterion_tests, common_nn.NewCriterionTest),
+]:
+    module_impl_check.add_tests(TestCppApiParity, test_params_dicts, test_instance_class, parity_table, devices)
```
Inside each of these, there is a check 
```
if 'FunctionalModule' in str(test_params_dict.get('constructor', '')):
            continue
```
that makes it so that for each item, the item only gets added to one of `module_impl_check` and `functional_impl_check`. Could we put that check here to avoid duplicating the check?

yf225 replied:
Thanks for the catch! Fixed :D

zou3519 commented on test/cpp_api_parity/module_impl_check.py:
```diff
@@ -0,0 +1,293 @@
+            unit_test_class.assertTrue(
+                key in cpp_grad_dict,
+                generate_error_msg("\"Does module have a parameter named `{}`?\"".format(key[:-5]), False, True))
+            unit_test_class.assertTrue(
```
Is there a reason we're using `assertTrue(torch.allclose(A, B), errmsg)` instead of `assertEqual(A, B, msg=errmsg)`?

yf225 replied:
Changed to use `assertEqual(A, B, message=errmsg)` :D

glaringlee commented on test/cpp_api_parity/module_impl_check.py:
```diff
@@ -0,0 +1,293 @@
+    test_params_dict['constructor'] = test_params_dict.get('constructor', getattr(torch.nn, module_name))
+    test_instance = test_instance_class(**test_params_dict)
+    assert test_instance.get_name().startswith('test_')
+    module_variant_name = test_instance.get_name()[5:] + (('_' + device) if device != 'cpu' else '')
```
I see multiple places that we concatenate multiple strings to form a variable name, sometimes use hardcoded stuff as well. Can we put a comment here or somewhere to give a full name sample or example so it won't be hard to trace what is this name look like.

yf225 replied:
Thanks for the catch! I added example outputs for `module_variant_name` and `functional_variant_name`. Please let me know if other examples are needed and I can add them in as well :D

glaringlee replied:
thx Will

glaringlee commented on test/cpp_api_parity/utils.py:
```diff
@@ -0,0 +1,247 @@
+import torch.utils.cpp_extension
+import torch.testing._internal.common_nn as common_nn
+from torch.testing._internal.common_cuda import TEST_CUDA
+
```
suggestion, It would be good if we can put a short description for the following 2 namedtuples, what is the usage of each variable inside the tuples. We do have samples to demonstrate the usage of these variables, but it did not cover all of these, for eg 'cpp_function_call', and it would same a lot of time for people who want to add new test cases.

yf225 replied:
Done :D The main note on this is `NOTE [How to check NN module / functional API parity between Python and C++ frontends]` in common_nn.py

Commit:
Update on "Refactor C++ API parity test mechanism and turn it on in CI again"

Turn on C++ API parity test again

gh-metadata: pytorch pytorch 35190 gh/yf225/106/head

Commit:
Update on "Refactor C++ API parity test mechanism and turn it on in CI again"

Turn on C++ API parity test again

gh-metadata: pytorch pytorch 35190 gh/yf225/106/head

Commit:
Update on "Refactor C++ API parity test mechanism and turn it on in CI again"

Turn on C++ API parity test again

gh-metadata: pytorch pytorch 35190 gh/yf225/106/head

Commit:
Update on "Refactor C++ API parity test mechanism and turn it on in CI again"

Turn on C++ API parity test again

gh-metadata: pytorch pytorch 35190 gh/yf225/106/head

Commit:
Update on "Refactor C++ API parity test mechanism and turn it on in CI again"

Turn on C++ API parity test again

gh-metadata: pytorch pytorch 35190 gh/yf225/106/head

Commit:
Update on "Refactor C++ API parity test mechanism and turn it on in CI again"

Turn on C++ API parity test again

gh-metadata: pytorch pytorch 35190 gh/yf225/106/head

Commit:
Update on "Refactor C++ API parity test mechanism and turn it on in CI again"

Turn on C++ API parity test again

gh-metadata: pytorch pytorch 35190 gh/yf225/106/head

yf225 commented:
@zou3519 @glaringlee Thanks so much for the review comments! This PR is ready for review again, and please let me know if anything is still unclear

Commit:
Update on "Refactor C++ API parity test mechanism and turn it on in CI again"

Turn on C++ API parity test again

gh-metadata: pytorch pytorch 35190 gh/yf225/106/head

Commit:
Update on "Refactor C++ API parity test mechanism and turn it on in CI again"

Turn on C++ API parity test again

gh-metadata: pytorch pytorch 35190 gh/yf225/106/head

glaringlee commented:
I saw a flake8-py3 check failure, are we supposed to fix that? or this is expected?

Commit:
Update on "Refactor C++ API parity test mechanism and turn it on in CI again"

Turn on C++ API parity test again

gh-metadata: pytorch pytorch 35190 gh/yf225/106/head

yf225 commented:
> I saw a flake8-py3 check failure, are we supposed to fix that? or this is expected?

Yes just fixed :)

zou3519 commented on torch/testing/_internal/common_nn.py:
```diff
@@ -364,7 +458,10 @@ def bce_with_logistic_no_reduce_test():
+        cpp_function_call='''F::binary_cross_entropy_with_logits(
+            i, t.to(i.options()), F::BinaryCrossEntropyWithLogitsFuncOptions().reduction(torch::kNone))''',
         input_fn=lambda: torch.rand(15, 10).clamp_(2.8e-2, 1 - 2.8e-2),
+        cpp_var_map={'i': 'input', 't': t},
```
It has bugged me a little while reading this that we are using the string "input" as a special value that means "ask the module test for inputs, and pass those in". On the other hand, I don't know what else would be better

yf225 replied:
I think an alternative is to require people to create the `input` tensor outside of the test dict, and do `input=the_input_tensor` in the test dict. But I am not sure if that would impose too much restriction on how people write the tests

zou3519 replied:
I think I would feel better about the magic here if we changed the magic string to make it more obvious that it's a magic string. Something like 'get_input()'

zou3519 replied:
Or if we passed in a global from somewhere, cpp_var_map={'i': GET_INPUT, 't': t},

yf225 replied:
Thanks for the suggestions! I changed it to `"_get_input()"`, to look exactly like the method we are calling on the test instance

zou3519 commented on test/test_cpp_api_parity.py:
```diff
@@ -1,687 +1,54 @@
+
+expected_test_params_dicts = []
+
+for test_params_dicts, test_instance_class in [
```
I think the following logic will make it more explicit that we're not accidentally duplicating the tests. I know that we have some asserts over the place to check this, but it's not entirely obvious from a developer's perspective because those asserts are elsewhere and not here
```
for test_params_dict, test_instance_class in [
    ...
]:
    for test_params_dict in test_params_dict:
        if is_functional_test(test_params_dict):
            functional_impl_check.write_test(TestCppApiParity, test_param_dict, test_instance_class, parity_table, devices)
        else:
             module_impl_check.write_test(TestCppApiParity, ...)
```

zou3519 replied:
Oh, the assert is actually on the next line lol.

zou3519 commented on test/cpp_api_parity/functional_impl_check.py:
```diff
@@ -0,0 +1,258 @@
+        cpp_tmp_folder=tempfile.mkdtemp(),
+    )
+
+def test_torch_nn_functional_variant(unit_test_class, test_params):
```
nit: Single line function seems unnecessary

zou3519 commented on test/cpp_api_parity/module_impl_check.py:
```diff
@@ -0,0 +1,308 @@
+    except:  # noqa: E722
+        pass
+
+def test_torch_nn_module_variant(unit_test_class, test_params):
```
nit: single line function seems unnecessary?

zou3519 commented on test/cpp_api_parity/utils.py:
```diff
@@ -0,0 +1,332 @@
+        raise RuntimeError("Found two tests with the same name: " + test_name)
+    setattr(unit_test_class, test_name, test_fn)
+
+def set_cpp_tensors_requires_grad(cpp_tensor_stmts, cpp_tensors):
```
What exactly is `cpp_tensors`? From the zip statement below it doesn't look like a list of python `torch.Tensor`

yf225 replied:
It actually is a list of python `torch.Tensor` - I renamed the argument to `python_tensors` to make it more clear

zou3519 commented on test/cpp_api_parity/utils.py:
```diff
@@ -0,0 +1,332 @@
+    return isinstance(test_instance, common_nn.CriterionTest) or \
+        isinstance(test_instance, common_nn.NewCriterionTest)
+
+def compute_cpp_args_construction_stmts_and_forward_arg_symbols(test_params):
```
What exactly does this function do?

zou3519 replied:
nvm, I figured it out by reading the code templates

yf225 replied:
Added comments for this function :D

zou3519 replied:
thanks, these really improve the readability :)

zou3519 commented on test/cpp_api_parity/functional_impl_check.py:
```diff
@@ -0,0 +1,258 @@
+            cpp_sources += cpp_api_parity.functional_cpp_sources.get(test_params.functional_name, '')
+            functionals_added_cpp_sources.add(test_params.functional_name)
+        cpp_sources += generate_test_cpp_sources(test_params=test_params, template=TORCH_NN_FUNCTIONAL_TEST_FORWARD)
+        functions.append('{}_{}'.format(test_params.functional_variant_name, 'test_forward'))
```
nit: `functions.append('{}_test_forward_backward'.format(test_params.module_variant_name))`

zou3519 commented on test/cpp_api_parity/functional_impl_check.py:
```diff
@@ -0,0 +1,258 @@
+    assert len(unit_test_class.functional_test_params_map) > 0
+    cpp_sources = TORCH_NN_COMMON_TEST_HARNESS
+    functions = []
+    functionals_added_cpp_sources = set()
```
Why is this a set?

yf225 replied:
One functional can have multiple test dicts, and if we always add the functional's cpp source for each test dict, there would be redefinition issues when we are adding the 2nd test dict / 3rd test dict / etc., and that's why we want to use a set here to keep track of the functionals we already added cpp sources for

zou3519 replied:
Oh, I see. This is only done for `sample_functional`, right? If so it would be nice if we could separate out the logic (the sample functional tests vs general tests) because it's a bit confusing what is going on here

zou3519 replied:
I am thinking something like the following:
```
functionals_added_cpp_sources = [SAMPLE_FUNCTIONAL_CPP_SOURCE]
for test_name, test_params in unit_test_class.functional_test_params_map.items():
    cpp_sources += generate_test_cpp_sources(test_params=test_params, template=TORCH_NN_FUNCTIONAL_TEST_FORWARD)
    functions.append('{}_test_forward'.format(test_params.functional_variant_name))
    if print_cpp_source:
        print(cpp_sources)
```
This means that we don't need `cpp_api_parity.functional_cpp_sources['sample_functional'] = SAMPLE_FUNCTIONAL_CPP_SOURCE` anymore in `sample_functional.py`, and furthemore we don't need `cpp_api_parity.functional_cpp_sources`.

yf225 replied:
Thanks for the suggestion! Improved accordingly :D

zou3519 commented on test/cpp_api_parity/module_impl_check.py:
```diff
@@ -0,0 +1,308 @@
+    assert len(unit_test_class.module_test_params_map) > 0
+    cpp_sources = TORCH_NN_COMMON_TEST_HARNESS
+    functions = []
+    modules_added_cpp_sources = set()
```
Why is this a set?

yf225 replied:
One module can have multiple corresponding test dicts, and if we always add the module cpp source for each test dict, there would be redefinition issues when we are adding the 2nd test dict / 3rd test dict / etc., and that's why we want to use a set here to keep track of the modules we already added cpp sources for

zou3519 commented on test/cpp_api_parity/module_impl_check.py:
```diff
@@ -0,0 +1,308 @@
+            modules_added_cpp_sources.add(test_params.module_name)
+        cpp_sources += generate_test_cpp_sources(
+            test_params=test_params, template=TORCH_NN_MODULE_TEST_FORWARD_BACKWARD)
+        functions.append('{}_{}'.format(test_params.module_variant_name, 'test_forward_backward'))
```
nit: don't need to pass the second argument as a format

zou3519 commented on test/test_cpp_api_parity.py:
```diff
@@ -1,687 +1,54 @@
 
 if __name__ == "__main__":
+    module_impl_check.build_cpp_tests(TestCppApiParity, print_cpp_source=PRINT_CPP_SOURCE)
+    functional_impl_check.build_cpp_tests(TestCppApiParity, print_cpp_source=PRINT_CPP_SOURCE)
```
TestCppApiParity doesn't work with pytest. I'm not sure how important it is that that works.

yf225 replied:
I moved the two `build_cpp_tests` lines out from the `if __name__ == "__main__"` block, which makes `TestCppApiParity` work with pytest as well :D

zou3519 commented on test/cpp_api_parity/functional_impl_check.py:
```diff
@@ -0,0 +1,258 @@
+                test_fn=test_fn,
+                test_cpp_api_parity=test_params_dict.get('test_cpp_api_parity', True),
+                test_cuda=test_params_dict.get('test_cuda', True),
+                has_impl_parity=parity_table['torch::nn::functional'][functional_full_name][0] and
```
nit: this is pretty hard to read. It might be nicer if parity_table were a class with some nice methods on it to get things. for example,
```
parity_table.find('torch::nn::functional::' + functional_full_name).has_impl_parity
```
but that's not really necessary

yf225 replied:
Agreed that it would make it easier to read. We can do it in a follow-up PR

zou3519 commented on test/cpp_api_parity/functional_impl_check.py:
```diff
@@ -0,0 +1,258 @@
+def add_torch_nn_functional_impl_parity_tests(
+        parity_table, unit_test_class, test_params_dicts, test_instance_class, devices):
+    if not hasattr(unit_test_class, 'functional_test_params_map'):
+        unit_test_class.functional_test_params_map = {}
```
Can we just add an empty dict to the unit_test_class instead of checking every time?

zou3519 commented on test/cpp_api_parity/functional_impl_check.py:
```diff
@@ -0,0 +1,258 @@
+            cpp_sources += cpp_api_parity.functional_cpp_sources.get(test_params.functional_name, '')
+            functionals_added_cpp_sources.add(test_params.functional_name)
+        cpp_sources += generate_test_cpp_sources(test_params=test_params, template=TORCH_NN_FUNCTIONAL_TEST_FORWARD)
+        functions.append('{}_{}'.format(test_params.functional_variant_name, 'test_forward'))
```
Could we just add a `test_name` to the `test_params` named tuple?

yf225 replied:
Sounds great and I think we should do it, and I think we can do it at a later time when we add other test functions (such as `test_constructor_args` or `test_init`) for each module test dict.

zou3519 commented on test/cpp_api_parity/functional_impl_check.py:
```diff
@@ -0,0 +1,258 @@
+        forward_output_file_path = compute_temp_file_path(cpp_tmp_folder, functional_variant_name, 'forward_output')
+
+        cpp_test_fn(arg_dict_file_path, forward_output_file_path)
+        cpp_output = torch.load(forward_output_file_path)
```
nit: There's a lot of I/O going on with these tests, which increases the risk of flakiness. If that becomes a problem in the future maybe someone can use pybind to connect the result of the cpp api to python

yf225 replied:
Sounds great and I noted it in the follow-up items to do

zou3519 commented on test/cpp_api_parity/functional_impl_check.py:
```diff
@@ -0,0 +1,258 @@
+
+    inputs = set_python_tensors_requires_grad(move_python_tensors_to_device(
+        [arg_value for _, arg_value in test_params.arg_dict['input']], device))
+    inputs = inputs + move_python_tensors_to_device(
```
nit: `inputs += blah`?

zou3519 commented on test/cpp_api_parity/functional_impl_check.py:
```diff
@@ -0,0 +1,258 @@
+
+    return python_output
+
+def test_forward(unit_test_class, test_params):
```
We should probably try to remove the cpp_tmp_folder if it exists already

zou3519 commented on test/cpp_api_parity/functional_impl_check.py:
```diff
@@ -0,0 +1,258 @@
+    try:
+        shutil.rmtree(cpp_tmp_folder)
+    except:  # noqa: E722
+        pass
```
nit: a warning is probably nice here

zou3519 commented on test/cpp_api_parity/module_impl_check.py:
```diff
@@ -0,0 +1,308 @@
+
+        assert hasattr(torch.nn, module_name), \
+            "`torch.nn` doesn't have module `{}`. ".format(module_name) + \
+            "If you are adding a new test, please set `fullname` using format `ModuleName_desc`, " + \
```
We should beef this up a little. The question I have while reading this is, "where am I setting these"? Is it in the module test dictionary?

zou3519 commented on test/cpp_api_parity/module_impl_check.py:
```diff
@@ -0,0 +1,308 @@
+        for key in python_grad_dict:
+            unit_test_class.assertTrue(
+                key in cpp_grad_dict,
+                msg=generate_error_msg("\"Does module have a parameter named `{}`?\"".format(key[:-5]), False, True))
```
nit: Should comment on where the -5 is coming from. This would probably be nice to do via asserting that all keys are in the form `grad_*`

zou3519 commented on test/cpp_api_parity/module_impl_check.py:
```diff
@@ -0,0 +1,308 @@
+
+def test_forward_backward(unit_test_class, test_params):
+    module_variant_name = test_params.module_variant_name
+    cpp_tmp_folder = test_params.cpp_tmp_folder
```
Nit: it's nice to remove the cpp_tmp_folder so that we can be more sure that a previous run of the test didn't affect the current run

Commit:
Update on "Refactor C++ API parity test mechanism and turn it on in CI again"

Turn on C++ API parity test again

gh-metadata: pytorch pytorch 35190 gh/yf225/106/head

Commit:
Update on "Refactor C++ API parity test mechanism and turn it on in CI again"

Turn on C++ API parity test again

gh-metadata: pytorch pytorch 35190 gh/yf225/106/head

Commit:
Update on "Refactor C++ API parity test mechanism and turn it on in CI again"

Turn on C++ API parity test again

gh-metadata: pytorch pytorch 35190 gh/yf225/106/head

Commit:
Update on "Refactor C++ API parity test mechanism and turn it on in CI again"

Turn on C++ API parity test again

gh-metadata: pytorch pytorch 35190 gh/yf225/106/head

Commit:
Update on "Refactor C++ API parity test mechanism and turn it on in CI again"

Turn on C++ API parity test again

gh-metadata: pytorch pytorch 35190 gh/yf225/106/head

zou3519 commented on test/cpp_api_parity/functional_impl_check.py:
```diff
@@ -0,0 +1,249 @@
+
+        add_test(unit_test_class, unit_test_name, test_fn)
+
+def write_test_to_test_class(unit_test_class, test_params_dict, test_instance_class, parity_table, devices):
```
nit: single line function seems unnecessary

yf225 replied:
Fixed :D

zou3519 commented on test/cpp_api_parity/module_impl_check.py:
```diff
@@ -0,0 +1,310 @@
+
+        add_test(unit_test_class, unit_test_name, test_fn)
+
+def write_test_to_test_class(unit_test_class, test_params_dict, test_instance_class, parity_table, devices):
```
nit: single line function seems unnecessary?

yf225 replied:
Fixed :D

zou3519 commented on test/cpp_api_parity/functional_impl_check.py:
```diff
@@ -0,0 +1,249 @@
+}
+""")
+
+def run_forward(unit_test_class, test_params):
```
So, just making sure I understand: the C++ API parity test mechanism re-implements running the forward + backward of a module test. The Module tests in test_nn.py still run using their own, separate mechanism.

Is there a plan in the future to unify the two mechanisms? Afaict the C++ API parity test mechanism doesn't support some things that the test_nn.py mechanism does; off of the top of my head `check_gradgrad` (which defaults to True) or precision checks (`check_half`) don't look supported

zou3519 replied:
(this is not blocking, I'm just curious)

yf225 replied:
That's correct - the C++ API parity mechanism and module tests in `test_nn.py` are separate now. Agreed that it should be possible to merge the two, and as a result expand the tests we run on the C++ modules

Commit:
Update on "Refactor C++ API parity test mechanism and turn it on in CI again"

Turn on C++ API parity test again

gh-metadata: pytorch pytorch 35190 gh/yf225/106/head

Commit:
Update on "Refactor C++ API parity test mechanism and turn it on in CI again"

Turn on C++ API parity test again

gh-metadata: pytorch pytorch 35190 gh/yf225/106/head

zou3519 reviewed (approved):
lgtm
